File: docs/Drasil/src/swift/InputConstraints.swift
/** InputConstraints.swift
    Provides the function for checking the physical constraints and software constraints on the input
    - Authors: Ting-Yu Wu
*/
import Foundation

/** Verifies that input values satisfy the physical constraints and software constraints
    - Parameter inParams: structure holding the input values
    - Parameter pi: ratio of circumference to diameter for any circle
*/
func input_constraints(_ inParams: inout InputParameters, _ pi: Double) throws -> Void {
    var outfile: FileHandle
    do {
        outfile = try FileHandle(forWritingTo: FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent("log.txt"))
        try outfile.seekToEnd()
    } catch {
        throw "Error opening file."
    }
    do {
        try outfile.write(contentsOf: Data("function input_constraints called with inputs: {".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  inParams = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("Instance of InputParameters object".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(", ".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  pi = ".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data(String(pi).utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.write(contentsOf: Data("  }".utf8))
        try outfile.write(contentsOf: Data("\n".utf8))
    } catch {
        throw "Error printing to file."
    }
    do {
        try outfile.close()
    } catch {
        throw "Error closing file."
    }
    
    if !(Double(-100000) <= inParams.F_vect_1 && inParams.F_vect_1 <= Double(100000)) {
        print("F_vect_1 has value ", terminator: "")
        print(inParams.F_vect_1, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("between ", terminator: "")
        print(-100000, terminator: "")
        print(" (F_vect_min)", terminator: "")
        print(" and ", terminator: "")
        print(100000, terminator: "")
        print(" (F_vect_max)", terminator: "")
        print(".")
        throw "InputError"
    }
    if !(0.0 < inParams.x_1 && inParams.x_1 <= Double(100000)) {
        print("x_1 has value ", terminator: "")
        print(inParams.x_1, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("between ", terminator: "")
        print(0.0, terminator: "")
        print(" (d_min)", terminator: "")
        print(" and ", terminator: "")
        print(100000, terminator: "")
        print(" (d_max)", terminator: "")
        print(".")
        throw "InputError"
    }
    if !(0.0 < inParams.x_2 && inParams.x_2 <= Double(100000)) {
        print("x_2 has value ", terminator: "")
        print(inParams.x_2, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("between ", terminator: "")
        print(0.0, terminator: "")
        print(" (d_min)", terminator: "")
        print(" and ", terminator: "")
        print(100000, terminator: "")
        print(" (d_max)", terminator: "")
        print(".")
        throw "InputError"
    }
    if !(pi / Double(2) < inParams.θ_1 && inParams.θ_1 < pi / Double(2)) {
        print("θ_1 has value ", terminator: "")
        print(inParams.θ_1, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("between ", terminator: "")
        print(pi / Double(2), terminator: "")
        print(" (θ_min)", terminator: "")
        print(" and ", terminator: "")
        print(pi / Double(2), terminator: "")
        print(" (θ_max)", terminator: "")
        print(".")
        throw "InputError"
    }
    if !(pi / Double(2) < inParams.θ_2 && inParams.θ_2 < pi / Double(2)) {
        print("θ_2 has value ", terminator: "")
        print(inParams.θ_2, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("between ", terminator: "")
        print(pi / Double(2), terminator: "")
        print(" (θ_min)", terminator: "")
        print(" and ", terminator: "")
        print(pi / Double(2), terminator: "")
        print(" (θ_max)", terminator: "")
        print(".")
        throw "InputError"
    }
    
    if !(inParams.F_vect_1 > Double(0)) {
        print("F_vect_1 has value ", terminator: "")
        print(inParams.F_vect_1, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("above ", terminator: "")
        print(0, terminator: "")
        print(".")
        throw "InputError"
    }
    if !(inParams.x_1 > Double(0)) {
        print("x_1 has value ", terminator: "")
        print(inParams.x_1, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("above ", terminator: "")
        print(0, terminator: "")
        print(".")
        throw "InputError"
    }
    if !(inParams.x_2 > Double(0)) {
        print("x_2 has value ", terminator: "")
        print(inParams.x_2, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("above ", terminator: "")
        print(0, terminator: "")
        print(".")
        throw "InputError"
    }
    if !(Double(0) < inParams.θ_1 && inParams.θ_1 < pi) {
        print("θ_1 has value ", terminator: "")
        print(inParams.θ_1, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("between ", terminator: "")
        print(0, terminator: "")
        print(" and ", terminator: "")
        print(pi, terminator: "")
        print(" (pi)", terminator: "")
        print(".")
        throw "InputError"
    }
    if !(Double(0) < inParams.θ_2 && inParams.θ_2 < pi) {
        print("θ_2 has value ", terminator: "")
        print(inParams.θ_2, terminator: "")
        print(", but is expected to be ", terminator: "")
        print("between ", terminator: "")
        print(0, terminator: "")
        print(" and ", terminator: "")
        print(pi, terminator: "")
        print(" (pi)", terminator: "")
        print(".")
        throw "InputError"
    }
}

File: docs/Drasil/src/csharp/InputConstraints.cs
/** \file InputConstraints.cs
    \author Ting-Yu Wu
    \brief Provides the function for checking the physical constraints and software constraints on the input
*/
using System;
using System.IO;

public class InputConstraints {
    
    /** \brief Verifies that input values satisfy the physical constraints and software constraints
        \param inParams structure holding the input values
        \param pi ratio of circumference to diameter for any circle
    */
    public static void input_constraints(InputParameters inParams, double pi) {
        StreamWriter outfile;
        outfile = new StreamWriter("log.txt", true);
        outfile.WriteLine("function input_constraints called with inputs: {");
        outfile.Write("  inParams = ");
        outfile.Write("Instance of InputParameters object");
        outfile.WriteLine(", ");
        outfile.Write("  pi = ");
        outfile.WriteLine(pi);
        outfile.WriteLine("  }");
        outfile.Close();
        
        if (!(-100000 <= inParams.F_vect_1 && inParams.F_vect_1 <= 100000)) {
            Console.Write("F_vect_1 has value ");
            Console.Write(inParams.F_vect_1);
            Console.Write(", but is expected to be ");
            Console.Write("between ");
            Console.Write(-100000);
            Console.Write(" (F_vect_min)");
            Console.Write(" and ");
            Console.Write(100000);
            Console.Write(" (F_vect_max)");
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
        if (!(0.0 < inParams.x_1 && inParams.x_1 <= 100000)) {
            Console.Write("x_1 has value ");
            Console.Write(inParams.x_1);
            Console.Write(", but is expected to be ");
            Console.Write("between ");
            Console.Write(0.0);
            Console.Write(" (d_min)");
            Console.Write(" and ");
            Console.Write(100000);
            Console.Write(" (d_max)");
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
        if (!(0.0 < inParams.x_2 && inParams.x_2 <= 100000)) {
            Console.Write("x_2 has value ");
            Console.Write(inParams.x_2);
            Console.Write(", but is expected to be ");
            Console.Write("between ");
            Console.Write(0.0);
            Console.Write(" (d_min)");
            Console.Write(" and ");
            Console.Write(100000);
            Console.Write(" (d_max)");
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
        if (!(pi / 2 < inParams.θ_1 && inParams.θ_1 < pi / 2)) {
            Console.Write("θ_1 has value ");
            Console.Write(inParams.θ_1);
            Console.Write(", but is expected to be ");
            Console.Write("between ");
            Console.Write(pi / 2);
            Console.Write(" (θ_min)");
            Console.Write(" and ");
            Console.Write(pi / 2);
            Console.Write(" (θ_max)");
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
        if (!(pi / 2 < inParams.θ_2 && inParams.θ_2 < pi / 2)) {
            Console.Write("θ_2 has value ");
            Console.Write(inParams.θ_2);
            Console.Write(", but is expected to be ");
            Console.Write("between ");
            Console.Write(pi / 2);
            Console.Write(" (θ_min)");
            Console.Write(" and ");
            Console.Write(pi / 2);
            Console.Write(" (θ_max)");
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
        
        if (!(inParams.F_vect_1 > 0)) {
            Console.Write("F_vect_1 has value ");
            Console.Write(inParams.F_vect_1);
            Console.Write(", but is expected to be ");
            Console.Write("above ");
            Console.Write(0);
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
        if (!(inParams.x_1 > 0)) {
            Console.Write("x_1 has value ");
            Console.Write(inParams.x_1);
            Console.Write(", but is expected to be ");
            Console.Write("above ");
            Console.Write(0);
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
        if (!(inParams.x_2 > 0)) {
            Console.Write("x_2 has value ");
            Console.Write(inParams.x_2);
            Console.Write(", but is expected to be ");
            Console.Write("above ");
            Console.Write(0);
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
        if (!(0 < inParams.θ_1 && inParams.θ_1 < pi)) {
            Console.Write("θ_1 has value ");
            Console.Write(inParams.θ_1);
            Console.Write(", but is expected to be ");
            Console.Write("between ");
            Console.Write(0);
            Console.Write(" and ");
            Console.Write(pi);
            Console.Write(" (pi)");
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
        if (!(0 < inParams.θ_2 && inParams.θ_2 < pi)) {
            Console.Write("θ_2 has value ");
            Console.Write(inParams.θ_2);
            Console.Write(", but is expected to be ");
            Console.Write("between ");
            Console.Write(0);
            Console.Write(" and ");
            Console.Write(pi);
            Console.Write(" (pi)");
            Console.WriteLine(".");
            throw new Exception("InputError");
        }
    }
}
